### DISCUSSION TITLE
Batch-adding excersises

### DISCUSSION THREAD

**Discussion by erikderzweite** [wger-project/wger#1621] [Q&A] [answered]
I have a Finnlo BioForce workout device and I want to add all the exercises for it to my wger instance. I have extracted the needed information out of instruction PDFs, created a database with relevant categories, equipment and pictures.
As I, of course, do not want to add over a hundred exercises manually, I've tried to add them with SQL, I have generated the needed statements like the above:

`INSERT INTO exercises_equipment VALUES ( '12', 'Fußschlaufen');
INSERT INTO exercises_equipment VALUES ( '13', 'Handgriffe');
INSERT INTO exercises_equipment VALUES ( '14', 'Latzugstange');
INSERT INTO exercises_equipment VALUES ( '15', 'Sitz');
INSERT INTO exercises_equipment VALUES ( '16', 'Beinstrecker');
INSERT INTO exercises_exercisebase VALUES ('3001', 'HAMMER SPORT AG', '9', 2, NULL, 'ae83749b-273b-ff12-7814-56fea441efe3', '2023-07-30', '2023-07-30 19:07:52.205965+0200', '', '', '', '' );
INSERT INTO exercises_exercise VALUES (3001, 'HAMMER SPORT AG', '<p>START: Stehen Sie mit sicherem Stand zum Gerät hin. Halten Sie die unteren Handgriffe auf Schulterhöhe. </p><p>BEWEGUNG: In die Knie gehen, bis Oberschenkel parallel zum Boden sind. Anspannung kurz halten und zur Ausgangsposition zurückkehren. </p><p>TIPPS: Knie auf gleicher Höhe mit den Füßen halten. Gewicht auf den ganzen Fuß verteilen, nicht nur auf Zehen oder Fersen. Oberkörper während der kompletten Übung gerade halten. </p><p>Schwierigkeit: Anfänger </p>', 'KNIEBEUGEN ', '2023-07-30', '1', 2, 'ae83749b-273b-ff12-7814-56fea441efe3', '3001', '2023-07-30 19:07:52.205965+0200', '', '', '', '' );
INSERT INTO exercises_exerciseimage VALUES (400, 'HAMMER SPORT AG', 'exercise-images/renamed_images/44-1.gif', 't', '3001', 2, '5c43079b-13e2-4744-593b-d172758fa803', 4 , '', '', '', '');
INSERT INTO exercises_exerciseimage VALUES (401, 'HAMMER SPORT AG', 'exercise-images/renamed_images/44-1-pulleys.png', 'f',  '3001', 2, '9affbec5-e651-fdaf-5729-cba0b73c8bb9', 4 , '', '', '', '');
INSERT INTO exercises_exerciseimage VALUES (402, 'HAMMER SPORT AG', 'exercise-images/renamed_images/44-1-1.png', 'f',  '3001', 2, '42389e22-547b-8797-4612-ce9d5168cdb5', 4 , '', '', '', '');
INSERT INTO exercises_exerciseimage VALUES (403, 'HAMMER SPORT AG', 'exercise-images/renamed_images/44-1-2.png', 'f',  '3001', 2, 'f66a37fd-a774-c1b7-86b1-32a5760cb9e2', 4 , '', '', '', '');
INSERT INTO exercises_exercisebase_equipment VALUES ('500', '3001', '11' );
INSERT INTO exercises_exercisebase_equipment VALUES ('501', '3001', '13' );
INSERT INTO exercises_exercisebase_muscles VALUES ('600', '3001', '10' );
INSERT INTO exercises_exercisebase_muscles VALUES ('601', '3001', '8' );
INSERT INTO exercises_exercisebase_muscles VALUES ('602', '3001', '7' );

INSERT INTO exercises_exercise VALUES (3002, 'HAMMER SPORT AG', '<p>START:  Stand facing frame. Feet hip width apart on base. Grip handles at shoulders. </p><p>Motion:  Flex knees, hips and ankles. Lower until thighs are parallel to floor or slightly higher. Press up to starting position. </p><p>TIPS:  Keep knees in line with toes. Center weight over arches, not toes or heels. Lift chest and keep back straight as hips lower. Keep head lifted and in line with spine. Contract abdominals to stabilize back. </p><p>DIFFICULTY:  Beginner </p>', 'SQUAT ', '2023-07-30', 2, 2, 'f9f74dec-9ae5-7b55-69ed-e7977657baa4', '3001', '2023-07-30 19:07:52.205965+0200', '', '', '', '' );`

It works in the web interface (I can't add or remove pictures though), but not in the android app: the overview page for such an exercise is grey. If I create same exercise manually, then the overview page works in android as well.

 How do I add a bunch of exercises to Wger? Is it possible via REST API?

**Comment by erikderzweite** [comment]
I try to add an exercise with REST API via the Swagger interface. I post to exercise-base the following:
`{
  "category": 9,
  "muscles": [10],
  "muscles_secondary": [],
  "equipment": [16, 11, 15],
  "license_author": "HAMMER SPORT AG",
  "variations": null
}`
get 201 as a response:
`{
  "id": 1025,
  "uuid": "01d76c73-e092-4b80-83f2-cf979ff74d39",
  "created": "2024-04-04T14:49:40.463148+02:00",
  "last_update": "2024-04-04T14:49:40.463189+02:00",
  "category": 9,
  "muscles": [
    10
  ],
  "muscles_secondary": [],
  "equipment": [
    16,
    11,
    15
  ],
  "variations": null,
  "license_author": "HAMMER SPORT AG"
}`

So far so good, 1025 is the id I get. I go to /api/v2/exercisebaseinfo/1025/, see 200 OK and the relevant data. If I search /api/v2/exercisebaseinfo/ for relevant category and equipment, the 1025 is in the results. But the /api/v2/exercise-base/1025/ returns 404 and searching exercise-base does not return 1025 as a result. Why?

**Comment by rolandgeider** [comment]
Hi! Did you also add one translation over the API? There might be some caching involved, but if exercisebaseinfo returns something, exercise-base should do as well.

What do you get when you run `python manage.py exercises-health-check` ?

**Comment by rolandgeider** [comment]
But yeah, adding an exercise consists of adding a base with the necessary categories etc and then adding a translation to it, nothing more, so I'm unsure why it doesn't work for you

**Comment by erikderzweite** [comment]
Ok, did not know about the translation, thought I could GET the exercise straight after adding the base. After I've added it, I can see the exersice. Still, the android app yields gray screen when I start the workout with it. Need to look further.

This is an example translation POST:
`{
  "name": "BEINSTRECKER_test",
  "exercise_base": 1027,
  "description": "START: Sitzen Sie auf dem Gerät und klemmen Sie Ihre Beine zwischen die Polster der Beinstreckereinheit.",
  "language": 1,
  "license_author": "HAMMER SPORT AG"
}`

**Comment by rolandgeider** [comment]
try doing a `python manage.py warmup-exercise-api-cache --force` afterwards (but this shouldn't be necessary). Also there is a local cache on the app, go to the settings and clear it, and reload the app just to be on the safe side

**Comment by erikderzweite** [comment]
Thanks for the hint! The SQL batch method is still grey, so I'll go with API instead.

Is there a documented way to add an image? I tried

`		with open(image_1, 'rb') as image_file:
			binary_content = image_file.read()

		binary_string = base64.b64encode(binary_content).decode('utf-8')
		
		params_image = {"exercise_base": 1227,
						"image": binary_string,
						}

		r = requests.post(url=image_url, headers=headers, json=params_image)`

But get an error 
`{'image': ['The submitted data was not a file. Check the encoding type on the '
           'form.']}`

How do I submit an image via API? The documentation says only "string", but in which format?

**Reply by erikderzweite** [reply, marked as answer]
Figured it out, you need to use "file" headers (of course), e.g.
`files = {
		    'license': (None, ''),
		    'license_derivative_source_url': (None, ''),
		    'license_author': (None, ''),
		    'license_object_url': (None, ''),
		    'is_main': (None, 'true'),
		    'style': (None, '1'),
		    'exercise_base': (None, '1252'),
		    'license_author_url': (None, ''),
		    'license_title': (None, ''),
		    'image': ('44-1-1.png', open(image_2, 'rb'), 'image/png'),
		}`
